File: Sources/RedlineCore/ClaudeSetup.swift
// What Claude Code has been configured with, read from the same files it reads. The findings
// checks compare this against what the transcripts show was actually used.
//
// Nothing here is written to, and nothing outside ~/.claude and a project's own CLAUDE.md is
// opened. A configuration file that cannot be parsed yields nothing rather than a guess.
import Foundation

public enum ClaudeSetup {
    /// Servers named in ~/.claude.json, both globally and per project, plus settings.json
    /// if a server block is there. Names only; nothing about how they are launched.
    public static func mcpServers(home: URL? = nil) -> [String] {
        let root = home ?? RedlineHome.url
        var names = Set<String>()
        func collect(_ any: Any?) {
            guard let dict = any as? [String: Any] else { return }
            for key in dict.keys { names.insert(key) }
        }
        if let json = readJSON(root.appendingPathComponent(".claude.json")) {
            collect(json["mcpServers"])
            if let projects = json["projects"] as? [String: Any] {
                for (_, value) in projects {
                    collect((value as? [String: Any])?["mcpServers"])
                }
            }
        }
        if let json = readJSON(root.appendingPathComponent(".claude/settings.json")) {
            collect(json["mcpServers"])
        }
        return names.sorted()
    }

    /// Skill directories under ~/.claude/skills. A skill is a directory holding SKILL.md,
    /// so a stray file is not counted as one.
    public static func skills(home: URL? = nil) -> [String] {
        let root = (home ?? RedlineHome.url)
            .appendingPathComponent(".claude/skills")
        return (try? FileManager.default.contentsOfDirectory(at: root,
                                                             includingPropertiesForKeys: nil))?
            .filter {
                FileManager.default.fileExists(
                    atPath: $0.appendingPathComponent("SKILL.md").path)
            }
            .map { $0.lastPathComponent }
            .sorted() ?? []
    }

    public static func agents(home: URL? = nil) -> [String] {
        markdownNames(in: (home ?? RedlineHome.url)
            .appendingPathComponent(".claude/agents"))
    }

    /// Commands, including one level of namespace directories, which is how Claude Code
    /// groups them. The namespace is dropped: the transcript records the invoked name.
    public static func commands(home: URL? = nil) -> [String] {
        let root = (home ?? RedlineHome.url)
            .appendingPathComponent(".claude/commands")
        var out = markdownNames(in: root)
        let fm = FileManager.default
        for url in (try? fm.contentsOfDirectory(at: root,
                                                includingPropertiesForKeys: [.isDirectoryKey]))
            ?? [] {
            guard (try? url.resourceValues(forKeys: [.isDirectoryKey]))?.isDirectory == true
            else { continue }
            out += markdownNames(in: url)
        }
        return Array(Set(out)).sorted()
    }

    static func markdownNames(in dir: URL) -> [String] {
        (try? FileManager.default.contentsOfDirectory(at: dir,
                                                      includingPropertiesForKeys: nil))?
            .filter { $0.pathExtension == "md" }
            .map { $0.deletingPathExtension().lastPathComponent }
            .sorted() ?? []
    }

    /// Memory files and their size in characters, with `@imports` expanded. Each file is
    /// counted once however many times it is imported, because the prompt holds it once.
    ///
    /// `sessionsByDir` says how many scanned sessions ran in each project directory, so the
    /// global file can be charged to every session and a project's file only to its own.
    public static func memoryFiles(sessionsByDir: [String: Int], home: URL? = nil)
        -> [MemoryFile] {
        let root = home ?? RedlineHome.url
        let total = sessionsByDir.values.reduce(0, +)
        var targets: [(url: URL, sessions: Int, global: Bool)] = [
            (root.appendingPathComponent(".claude/CLAUDE.md"), total, true),
        ]
        targets += sessionsByDir.keys.sorted().map {
            (URL(fileURLWithPath: $0).appendingPathComponent("CLAUDE.md"),
             sessionsByDir[$0] ?? 0, false)
        }
        var out: [MemoryFile] = []
        var seen = Set<String>()
        for target in targets {
            var chars = 0
            expand(target.url, home: root, depth: 0, seen: &seen) { chars += $0 }
            guard chars > 0 else { continue }
            out.append(MemoryFile(path: target.url.path, chars: chars,
                                  sessions: target.sessions, global: target.global))
        }
        return out
    }

    /// Depth is capped rather than cycle-detected as well: `seen` already prevents a loop,
    /// and a chain deeper than this is not something to silently follow.
    static func expand(_ url: URL, home: URL, depth: Int, seen: inout Set<String>,
                       add: (Int) -> Void) {
        guard depth < 3, !seen.contains(url.path),
              let text = try? String(contentsOf: url, encoding: .utf8) else { return }
        seen.insert(url.path)
        add(text.count)
        var imports: [URL] = []
        text.enumerateLines { line, _ in
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            guard trimmed.hasPrefix("@"), trimmed.count > 1 else { return }
            let raw = String(trimmed.dropFirst()).split(separator: " ").first.map(String.init)
            guard let raw, !raw.isEmpty else { return }
            if raw.hasPrefix("~/") {
                imports.append(home.appendingPathComponent(String(raw.dropFirst(2))))
            } else if AppPaths.isAbsolute(raw) {
                imports.append(URL(fileURLWithPath: raw))
            } else {
                imports.append(url.deletingLastPathComponent().appendingPathComponent(raw))
            }
        }
        for next in imports {
            expand(next, home: home, depth: depth + 1, seen: &seen, add: add)
        }
    }

    static func readJSON(_ url: URL) -> [String: Any]? {
        guard let data = try? Data(contentsOf: url) else { return nil }
        return try? JSONSerialization.jsonObject(with: data) as? [String: Any]
    }

    /// Gathers everything the checks need. Kept in one place so the app, the CLI and the
    /// tests all assemble the same input.
    public static func findingsInput(sessions: [SessionScan], windowDays: Int,
                                     now: Date = Date(), home: URL? = nil) -> FindingsInput {
        var byDir: [String: Int] = [:]
        for cwd in sessions.compactMap(\.cwd) { byDir[cwd, default: 0] += 1 }
        return FindingsInput(sessions: sessions,
                             configuredMCPServers: mcpServers(home: home),
                             skills: skills(home: home),
                             agents: agents(home: home),
                             commands: commands(home: home),
                             memoryFiles: memoryFiles(sessionsByDir: byDir, home: home),
                             windowDays: windowDays,
                             now: now)
    }
}

File: windows/src/Redline.Core/ClaudeSetup.cs
// What Claude Code has been configured with, read from the files it reads; nothing is written, and
// nothing outside ~/.claude and a project's own CLAUDE.md is opened. Unparseable yields nothing.
using System.Globalization;
using System.Text;
using System.Text.Json.Nodes;

namespace Redline.Core;

public static class ClaudeSetup
{
    /// Servers named in ~/.claude.json, globally and per project, plus settings.json. Names only.
    public static List<string> McpServers(string? home = null)
    {
        var root = home ?? RedlineHome.Url;
        var names = new HashSet<string>(StringComparer.Ordinal);
        void Collect(JsonNode? any)
        {
            if (any is not JsonObject dict) return;
            foreach (var (key, _) in dict) names.Add(key);
        }
        if (ReadJson(RedlineHome.Join(root, ".claude.json")) is { } json)
        {
            Collect(json["mcpServers"]);
            if (json["projects"] is JsonObject projects)
                foreach (var (_, value) in projects) Collect((value as JsonObject)?["mcpServers"]);
        }
        if (ReadJson(RedlineHome.Join(root, ".claude/settings.json")) is { } settings)
            Collect(settings["mcpServers"]);
        return names.OrderBy(n => n, StringComparer.Ordinal).ToList();
    }

    /// A skill is a directory under ~/.claude/skills holding SKILL.md; a stray file is not one.
    public static List<string> Skills(string? home = null)
    {
        var root = RedlineHome.Join(home ?? RedlineHome.Url, ".claude/skills");
        try
        {
            return Directory.GetFileSystemEntries(root)
                .Where(p => File.Exists(Path.Combine(p, "SKILL.md")))
                .Select(Path.GetFileName).OfType<string>()
                .OrderBy(n => n, StringComparer.Ordinal).ToList();
        }
        catch { return new(); }
    }

    public static List<string> Agents(string? home = null) =>
        MarkdownNames(RedlineHome.Join(home ?? RedlineHome.Url, ".claude/agents"));

    /// Commands, including one level of namespace directories. The namespace is dropped:
    /// the transcript records the invoked name.
    public static List<string> Commands(string? home = null)
    {
        var root = RedlineHome.Join(home ?? RedlineHome.Url, ".claude/commands");
        var output = MarkdownNames(root);
        try
        {
            foreach (var dir in Directory.GetDirectories(root)) output.AddRange(MarkdownNames(dir));
        }
        catch { }
        return output.Distinct(StringComparer.Ordinal).OrderBy(n => n, StringComparer.Ordinal).ToList();
    }

    internal static List<string> MarkdownNames(string dir)
    {
        try
        {
            return Directory.GetFileSystemEntries(dir)
                .Where(p => Path.GetExtension(p) == ".md")
                .Select(Path.GetFileNameWithoutExtension).OfType<string>()
                .OrderBy(n => n, StringComparer.Ordinal).ToList();
        }
        catch { return new(); }
    }

    /// Memory files and their size in characters with @imports expanded, each counted once.
    /// The global file is charged to every session, a project's only to its own.
    public static List<MemoryFile> MemoryFiles(IReadOnlyDictionary<string, int> sessionsByDir, string? home = null)
    {
        var root = home ?? RedlineHome.Url;
        var total = sessionsByDir.Values.Sum();
        var targets = new List<(string Path, int Sessions, bool Global)>
        {
            (RedlineHome.Join(root, ".claude/CLAUDE.md"), total, true),
        };
        targets.AddRange(sessionsByDir.Keys.OrderBy(k => k, StringComparer.Ordinal)
            .Select(k => (Path.Combine(k, "CLAUDE.md"), sessionsByDir[k], false)));
        var output = new List<MemoryFile>();
        // Windows paths compare without case, so one file reached two ways counts once
        var seen = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        foreach (var target in targets)
        {
            var chars = 0;
            Expand(target.Path, root, 0, seen, n => chars += n);
            if (chars <= 0) continue;
            output.Add(new MemoryFile(target.Path, chars, target.Sessions, target.Global));
        }
        return output;
    }

    private static readonly UTF8Encoding StrictUtf8 = new(false, true);

    /// Depth is capped as well as `seen` preventing a loop: a deeper chain is not silently followed.
    internal static void Expand(string path, string home, int depth, HashSet<string> seen, Action<int> add)
    {
        if (depth >= 3) return;
        string key;
        try { key = Path.GetFullPath(path); } catch { return; }
        if (seen.Contains(key)) return;
        string text;
        try { text = File.ReadAllText(path, StrictUtf8); } catch { return; }
        seen.Add(key);
        // Characters as Swift counts them: grapheme clusters, not UTF-16 units
        add(new StringInfo(text).LengthInTextElements);
        var imports = new List<string>();
        foreach (var line in text.Split('\n'))
        {
            var trimmed = line.TrimEnd('\r').Trim(' ', '\t');
            if (!trimmed.StartsWith('@') || trimmed.Length <= 1) continue;
            var raw = trimmed[1..].Split(' ', StringSplitOptions.RemoveEmptyEntries).FirstOrDefault();
            if (string.IsNullOrEmpty(raw)) continue;
            if (raw.StartsWith("~/")) imports.Add(RedlineHome.Join(home, raw[2..]));
            else if (raw.StartsWith('/') || Path.IsPathFullyQualified(raw)) imports.Add(raw);
            else imports.Add(Path.Combine(Path.GetDirectoryName(path) ?? "", raw));
        }
        foreach (var next in imports) Expand(next, home, depth + 1, seen, add);
    }

    internal static JsonObject? ReadJson(string path) => Json.ReadObject(path);

    /// Gathers everything the checks need, so the app, the CLI and the tests assemble the same input.
    public static global::Redline.Core.FindingsInput FindingsInput(IReadOnlyList<SessionScan> sessions, int windowDays,
                                                                     DateTimeOffset? now = null, string? home = null)
    {
        var byDir = new Dictionary<string, int>();
        foreach (var cwd in sessions.Select(s => s.Cwd).OfType<string>())
            byDir[cwd] = byDir.GetValueOrDefault(cwd) + 1;
        return new global::Redline.Core.FindingsInput(sessions.ToList(), McpServers(home), Skills(home), Agents(home),
                                 Commands(home), MemoryFiles(byDir, home), windowDays,
                                 now ?? DateTimeOffset.UtcNow);
    }
}
